std/random.d

```diff
@@ -1290,6 +1290,48 @@ alias Xorshift    = Xorshift128;                            /// ditto
     }
 }
 
+version (CRuntime_Bionic)
+    version = SecureARC4Random; // ChaCha20
+version (OSX)
+    version = SecureARC4Random; // AES
+version (OpenBSD)
+    version = SecureARC4Random; // ChaCha20
+version (NetBSD)
+    version = SecureARC4Random; // ChaCha20
+
+version (CRuntime_UClibc)
+    version = LegacyARC4Random; // ARC4
+version (FreeBSD)
+    version = LegacyARC4Random; // ARC4
+version (DragonFlyBSD)
+    version = LegacyARC4Random; // ARC4
+version (BSD)
+    version = LegacyARC4Random; // Unknown implementation
+
+// For the current purpose of unpredictableSeed the difference between
+// a secure arc4random implementation and a legacy implementation is
+// unimportant. The source code documents this distinction in case in the
+// future Phobos is altered to require cryptographically secure sources
+// of randomness, and also so other people reading this source code (as
+// Phobos is often looked to as an example of good D programming practices)
+// do not mistakenly use insecure versions of arc4random in contexts where
+// cryptographically secure sources of randomness are needed.
+
+// Performance note: ChaCha20 is about 70% faster than ARC4, contrary to
+// what one might assume from it being more secure.
+
+version (SecureARC4Random)
+    version = AnyARC4Random;
+version (LegacyARC4Random)
+    version = AnyARC4Random;
+
+version (AnyARC4Random)
+{
+    extern(C) private @nogc nothrow
+    {
+        uint arc4random() @safe;
```

Comment:
>Is there any reason why `arc4random` should not be exposed publicly on platforms that offer it, together with `arc4random_buf` and `arc4random_uniform` ... ?

This PR has been edited not to introduce any new public symbols because new public symbols require @andralex's approval. See discussion: https://github.com/dlang/phobos/pull/6267#discussion_r173754093

Reply:
(If we wanted to be more D-esque, the `arc4random_buf` implementation might be provided as `arc4random_buf(void[] buf)` and invoke the C API version underneath the hood; or that could be provided as an additional overload.)

Reply:
> This PR has been edited not to introduce any new public symbols because new public symbols require @andralex's approval. See discussion: #6267 (comment)

Fair enough, but note that this does not mean we cannot pursue the compromise outlined above: we'd just have to split between the changes to `unpredictableSeed` and the introduction of public `arc4random` functions.

In the event of serious hassle over the latter, we could always revisit the question of which platforms get the under-the-hood `arc4random` `unpredictableSeed`.

Reply:
[mir-random does something along those lines](https://github.com/libmir/mir-random/blob/6561a66bd8660bb75222364ed65ee6c79cd33bd6/source/mir/random/engine/package.d#L1037-L1042), but abstracts away the difference between platforms.
```d
size_t genRandomNonBlocking()(scope ubyte[] buffer) @nogc nothrow @trusted
{
    pragma(inline, true);
    return mir_random_genRandomNonBlocking(buffer.ptr, buffer.length);
}
```
`mir_random_genRandomNonBlocking` (the awkward name is for C linkage) could be calling `arc4random_buf`, or it could be using Linux's `getrandom` syscall, etc. I think that's the right kind of interface.

Reply:
Yep, introducing new public symbols is a tricky business.
And even if we get an approval, it should probably go to `std.experimental.random` as experience shows that it's really hard to get all use-cases right. 

> we wanted to be more D-esque,

There are a lot of things that could be done, e.g.  `ref ubyte[]` (and of course camelCase), but I doubt that a user is ever going to care about the underlying details. What they care about is:
- cross-platform
- blocking/non-blocking
- `@nogc` vs . GC
- entropy quality
- speed

Sometimes these concerns can be combined, but they don't want to special-case their code just because their code might be used on OSX too.

> Is there any reason why arc4random should not be exposed publicly on platforms that offer it,

See reasons above + it's really easy to do so yourself as a user if you really want to.

Reply:
>  introduction of public arc4random functions.

BTW declaration of `extern(C)` functions is done in DRuntime. Phobos is supposed to contain the cross-platform high-level abstractions.

Reply:
That's what I assumed, hence the remarks about not exposing them via phobos.